src/librustc_ast_passes/ast_validation.rs

```diff
@@ -794,10 +794,10 @@ impl<'a> Visitor<'a> for AstValidator<'a> {
                             .help("use `auto trait Trait {}` instead")
                             .emit();
                     }
-                    if let (Unsafe::Yes(span), ImplPolarity::Negative) = (unsafety, polarity) {
+                    if let (Unsafe::Yes(span), ImplPolarity::Negative(sp)) = (unsafety, polarity) {
                         struct_span_err!(
                             this.session,
-                            item.span,
```

Comment:
Could we use `sm.def_span(item.span)` for the primary span and then add `.span_label(nsp, "negative because of this")` where `::Negative(nsp)` (ideally rename the unsafe span to `usp` for clarity)? The goal would be to underline the whole implementation as the cause, but point specifically to the `!` as the cause of negativity.

Reply:
Also, consider extracting this to another method, passing the spans in, as it is growing somewhat large.

Reply:
> The goal would be to underline the whole implementation as the cause, but point specifically to the ! as the cause of negativity.

I considered doing so, but I didn't want to make the output too verbose.

Reply:
It doesn't feel like two span labels would be too verbose, particularly in the case of just pointing to the `!` (not this case, but I'd prefer some uniformity in the diagnostics).